models/qwen3/14b/decode_fwd.py

```
@@ -250,13 +288,26 @@
+# every one produces exactly one TASK_ID. fa_fused depends on those Scalars directly;
+# repeating a SPMD task-id per output tile adds duplicate graph edges only.
+
+# Fused QK-norm reduction alignment. The per-(KV head, batch) sum-of-squares emits a
+# col-major [rows, 1] tile; ptoas requires its column byte size (rows * sizeof(FP32))
+# to be 32B-aligned, i.e. rows a multiple of 8. Q already pads to Q_HEAD_PAD (16) rows;
+# K's single real row is zero-padded to K_RED_ROWS for the reduction (then row 0 kept).
+K_RED_ROWS = 8
+assert (Q_HEAD_PAD * 4) % 32 == 0, "Q QK-norm reduction rows (Q_HEAD_PAD) must be 32B-aligned"
+assert (K_RED_ROWS * 4) % 32 == 0, "K QK-norm reduction rows (K_RED_ROWS) must be 32B-aligned"
+
 # ── Scope 3a · out_proj (split-K × split-N, atomic-add into attn_proj_fp32) ──
```

Ruff (0.15.20)

[warning] Comment contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF003)